### PR TITLE
Add defaults in automatic schema generation

### PR DESCRIPTION
**Pull request by wmshort**
Partially addresses #193 .

Initial attempt at default parsing and generation. This PR is mainly intended to kick off review, and changes will almost certainly need to be made. But I need another set of eyeballs on it, to see if I am even on the right track!

### CODE REVIEW

**Comment by wmshort** [comment]
Thanks. One open question is handling defaults for Interval columns; how do we get from the postgres string to IntervalCustom?

**Comment by dantownsend** [comment]
@wmshort Yeah, that one looks really tricky. It doesn't look like the easiest thing to parse.

Screenshot 2021-09-22 at 21 26 32

I'm not the best with regex - it might be possible ... but a Python function might be better. Something like:

```python
date, time = default.rsplit(' ', 1)
hour, minute, second = [int(i) for i in time.split(':')]
```

Once we've parsed it, it's easy enough to create the `IntervalCustom` - we just need to instantiate it with the relevant days / hours etc.

**Comment by wmshort** [comment]
In the postgres documentation I am not seeing the version where digital times are mixed with phrasal years / months / days. The examples in the docs are all like:

5 weeks 4 days 3 hours 2 minutes 1 second

I think I have a workable regex now for this fully phrasal type, but does it need to handle both kinds? Thus equivalent:

5 weeks 4 days 3 hours ago
5 weeks 4 days 03:00:00 ago

I guess it can't hurt...

File: piccolo/apps/schema/commands/generate.py [hunk 1]
```diff
@@ -174,15 +174,17 @@ def get_table_with_name(self, name: str) -> t.Optional[t.Type[Table]]:
...
     Bytea: re.compile(r"'(.*)'::bytea$"),
-    DoublePrecision: re.compile(r"[+-]?([0-9]*[.])?[0-9]+"),
+    DoublePrecision: re.compile(r"([+-]?(?:[0-9]*[.])?[0-9]+)"),
     Varchar: re.compile(r"^'(.*)'::character varying$"),
     Date: re.compile(r"^((?:\d{4}-\d{2}-\d{2})|CURRENT_DATE)$"),
     Integer: re.compile(r"^(-?\d+)$"),
-    Interval: re.compile(r"^'(.*)'::interval"),
-    JSON: re.compile(r"^'(.*)'::json$"),
+    Interval: re.compile(
+        r"^(?:')?(?:(?:(?P<years>\d+) y(?:ear(?:s)?)?\b)|(?:(?P<months>\d+) m(?:onth(?:s)?)?\b)|(?:(?P<weeks>\d+) w(?:eek(?:s)?)?\b)|(?:(?P<days>\d+) d(?:ay(?:s)?)?\b)|(?:(?:(?:(?P<hours>\d+) h(?:our(?:s)?)?\b)|(?:(?P<minutes>\d+) m(?:inute(?:s)?)?\b)|(?:(?P<seconds>\d+) s(?:econd(?:s)?)?\b))|(?:(?P<digits>-?\d\d:\d\d:\d\d))?\b))+(?P<direction>ago)?(?:'::interval)?$"
```
Comment (dantownsend) [inline piccolo/apps/schema/commands/generate.py:182]: That's an impressive regex!

Comment (wmshort) [reply piccolo/apps/schema/commands/generate.py:182]: It might be overkill, but it works...

**Comment by wmshort** [comment]
One more thing: callables as defaults. The Serial column utilizes an internal postgres function, nextval, to evaluate the next in the sequence. This is not callable in the same way as, say, UUID4. Is it even meant to be settable by the user?

**Comment by dantownsend** [comment]
That's a good point - if the user hardcoded the default for a Serial column to something like `5`, I'm not sure if Postgres would reject that, but wouldn't be good.

**Comment by dantownsend** [comment]
With the default values in Piccolo, it tries to mirror them as much as possible, so it exists in Python and the database.

So with UUID, if someone bypassed the ORM and inserted straight into the database, it would still behave itself.

There are some situations where the default can only exist in the database (e.g. Serial), or in Python (some custom user function).

**Comment by wmshort** [comment]
> With the default values in Piccolo, it tries to mirror them as much as possible, so it exists in Python and the database.
> 
> So with UUID, if someone bypassed the ORM and inserted straight into the database, it would still behave itself.
> 
> There are some situations where the default can only exist in the database (e.g. Serial), or in Python (some custom user function).

So for Serial we probably just want to omit the reflection of the default (nextval etc) in the generated schema -- i.e. just output column_name = Serial() -- as the mechanism is all handled internally. Or...?

**Comment by dantownsend** [comment]
@wmshort That's right, yeah - if it's `Serial` either omit the default argument, or set it to `None` or `...`.

**Comment by dantownsend** [comment]
@wmshort It's looking great - the tests failed due to a type error. It's because when I originally wrote the code I didn't add a type annotation for `COLUMN_TYPE_MAP`, so it's my mistake.

This should fix it:

```python
COLUMN_TYPE_MAP: t.Dict[str, t.Type[Column]] = {
```

**Comment by wmshort** [comment]
Sorry about that last one, I had moved on to the decimal precision and forgot to remove extraneous code :) should be all set now, though.

**Pull request merged**